File: 273_integer_to_english_words.py
'''
## Problem 273: Integer to english words

## Author: Neha Doiphode
## Date:   08-16-2022

## Description:
    Convert a non-negative integer num to its English words representation.

## Examples:
    Example 1:
        Input: num = 123
        Output: "One Hundred Twenty Three"

    Example 2:
        Input: num = 12345
        Output: "Twelve Thousand Three Hundred Forty Five"

    Example 3:
        Input: num = 1234567
        Output: "One Million Two Hundred Thirty Four Thousand Five Hundred Sixty Seven"

## Constraints:
    0 <= num <= 231 - 1

## Time complexity: O(1), since we are processing in the group of triplets and we are only going to have
                          maximum until billions.

## Space complexity: O(1), Even if we are using recursion, we are always going to have fixed stack length
                           as we are only calling it over 3 digits.

'''
def get_input():
    print("Enter the number: ", end = "")
    number = int(input())
    print()
    return number

class Solution:
    def helper(self, number: int) -> str:
        below_20 = [" ", "One", "Two", "Three", "Four", "Five",
                    "Six", "Seven", "Eight", "Nine", "Ten",
                    "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen",
                    "Sixteen", "Seventeen", "Eighteen", "Nineteen", "Twenty"]

        tens = [" ", "Ten", "Twenty", "Thirty", "Forty", "Fifty",
                "Sixty", "Seventy", "Eighty", "Ninety"]

        if number == 0:
            return ""
        elif number >= 1 and number < 20:
            return below_20[number] + " "
        elif number < 100:
            return tens[int(number / 10)] + " " + self.helper(number % 10)
        else:
            return below_20[int(number / 100)] + " Hundred " + self.helper(number % 100)

    def numberToWords(self, num: int) -> str:
        extensions = [" ", "Thousand", "Million", "Billion"]
        if num == 0:
            return "Zero"

        result = ""
        i = 0
        while num > 0:
            if num % 1000 > 0:
                result = self.helper(num % 1000) + extensions[i] + " " + result

            num = int(num / 1000)
            i += 1

        return result.strip()

# Driver code
solution = Solution()
number = get_input()
print(f"Input number: {number}")
print(f"Integer to words conversion: {solution.numberToWords(number)}")

File: 273. Integer to English Words.java
// Time Complexity : O(1)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach:
// We can divide this problem into sub problems of triplets.
// we need to find the english version of the 3 digit number and add the thousands place.
// to get the three digit number(987) we used a helper function if the number > 99 , we get the 100th place by /100(9) from the below20 and append hunder and call helper on num%100 (87)
// if number is less than 20 we get it from below 20 array
// if number is less than 99 (87).. we get tht tens (8) from the tens array and append the helper on (num%10)(7)
class Solution {
    static String[] below_20 ={"","One ","Two ","Three ","Four ","Five ","Six ","Seven ",
      "Eight ","Nine ","Ten ","Eleven ","Twelve ","Thirteen ","Fourteen ","Fifteen ","Sixteen ",
      "Seventeen ","Eighteen ","Nineteen "};

    static String[] tens = {"", "Ten ","Twenty ","Thirty ","Forty ","Fifty ","Sixty ",
      "Seventy ","Eighty ","Ninety "};

    static String[] thousands ={"", "Thousand ", "Million ", "Billion "};
    
    public String numberToWords(int num) {
        if(num ==0) return "Zero";
        String result = "";
        int i=0;
        while(num>0)
        {
          int triplet = num%1000;
          if(triplet!=0)
          {
              result =helper(triplet) + thousands[i] + result;
          }
          i++;
          num/=1000;
        }
        return result.trim();
    } 

    private String helper(int num)
    {
      if(num<20)
          return below_20[num];
      else if(num<100)
          return tens[num/10] + helper(num%10);
      else 
          return below_20[num/100] +"Hundred "+ helper(num%100);
  }
}